### PR TITLE
Show warnings on the overview related to quota

### PR DESCRIPTION
**Pull request by jwforres**
Partially implements https://trello.com/c/mfeqgKvL

### CODE REVIEW

**Comment by jwforres** [comment]
@spadgett PTAL

File: app/scripts/controllers/overview.js [hunk 1]
```diff
@@ -385,6 +386,32 @@ angular.module('openshiftConsole')
       $scope.renderOptions.showLoading = !loaded && projectEmpty;
     };
 
+    var setGenericQuotaWarning = function() {
+      var isHidden = AlertMessageService.isAlertPermanentlyHidden("overview-quota-limit-reached", $scope.projectName);
+      if (!isHidden && QuotaService.isAnyQuotaExceeded($scope.quotas, $scope.clusterQuotas)) {
+        $scope.alerts['quotaExceeded'] = {
```
Comment (spadgett) [inline app/scripts/controllers/overview.js]: Will the alert will reappear on the next poll if I close it with the `x`? If so, does

```js
if ($scope.alerts.quotaExceeded) {
  return;
}
```

prevent that?

File: app/scripts/directives/deploymentDonut.js [hunk 2]
```diff
@@ -53,6 +56,23 @@ angular.module('openshiftConsole')
         $scope.$watchGroup(['limitRanges', 'hpa', 'project'], updateHPAWarnings);
         $scope.$watch('rc.spec.template.spec.containers', updateHPAWarnings, true);
 
+        var updateQuotaWarning = function() {
+          if (_.get($scope.rc, 'spec.replicas', 0) > _.get($scope.rc, 'status.replicas', 0)) {
```
Comment (spadgett) [inline app/scripts/directives/deploymentDonut.js]: RC `spec.replicas` defaults to 1 if not set.

https://github.com/spadgett/origin/blob/fc85c362316ba6c7cd01341da2a2f27c186bdc6b/vendor/k8s.io/client-go/1.4/pkg/api/v1/types.go#L2011-L2015

Comment (spadgett) [inline app/scripts/directives/deploymentDonut.js]: It looks like `spec.replicas` is required for replica sets and defaults to 1 for pet sets.

File: app/scripts/services/quota.js [hunk 3]
```diff
@@ -81,9 +84,14 @@ angular.module("openshiftConsole")
         'requests.cpu',
         'requests.memory',
         'limits.cpu',
-        'limits.memory'
+        'limits.memory',
+        'pods'
       ], function(type) {
         var q = quota.status.total || quota.status;
+        // Don't check 'pods'' quota if the resource is a pod, that will create duplicate warnings in combination with getQuotaAlerts
```
Comment (spadgett) [inline app/scripts/services/quota.js]: comment nit: extra `'` after the word pods

File: app/scripts/services/quota.js [hunk 4]
```diff
@@ -185,6 +193,23 @@ angular.module("openshiftConsole")
...
+            if (quotaKey === 'resourcequotas') {
+              return false;
+            }
+            return !isNil(q.hard[quotaKey]) && usageValue(q.hard[quotaKey]) <= usageValue(q.used[quotaKey]);
+          });
+        };
+        if (_.some(quotas, isExceeded)) {
+          return true;
+        }
+        return _.some(clusterQuotas, isExceeded);
```
Comment (spadgett) [inline app/scripts/services/quota.js]:
```js
return _.some(quotas, isExceeded) || _.some(clusterQuotas, isExceeded);
```

?

File: app/scripts/services/quota.js [hunk 5]
```diff
@@ -185,6 +193,23 @@ angular.module("openshiftConsole")
...
 
+    var isAnyQuotaExceeded = function(quotas, clusterQuotas) {
+        var isExceeded = function(quota) {
+          var q = quota.status.total || quota.status;
+          return _.some(q.hard, function(_rawVal, quotaKey) {
+            // We always ignore quota warnings about being out of resourcequotas since end users cant do anything about it
+            if (quotaKey === 'resourcequotas') {
+              return false;
+            }
+            return !isNil(q.hard[quotaKey]) && usageValue(q.hard[quotaKey]) <= usageValue(q.used[quotaKey]);
```
Comment (spadgett) [inline app/scripts/services/quota.js]: Are we able to use `_rawVal` in place of `q.hard[quotaKey]`?

File: app/views/browse/_replica-set-details.html [hunk 6]
```diff
@@ -5,6 +5,8 @@
         deployment-config="deploymentConfig"
         pods="podsForDeployment"
         hpa="autoscalers"
+        quotas="quotas"
+        cluster-quotas="clusterQuotas"
```
Comment (spadgett) [inline app/views/browse/_replica-set-details.html]: I didn't think we were polling these values in the replica set controller?

Comment (jwforres) [inline app/views/browse/_replica-set-details.html]: yeah i changed all the templates before i decided not to add the quota polling in the RS controller

**Comment by jwforres** [comment]
@spadgett updated with review feedback

**Comment by jwforres** [comment]
[merge]

**Pull request merged**